### DISCUSSION TITLE
Parentheses for languages that don't require it aside from execution ordering

### DISCUSSION THREAD

**Discussion by katawful** [tree-sitter/tree-sitter#1104] [Q&A] [answered]
The language I'm creating a parser for doesn't need parentheses for any statement or function, but they can be used for ordering or containing function calls. Example of what I mean:

```
if x > y && (x < z || x == z) // the parentheses encloses the OR call properly
let x = function (another function) // the second function takes the place of an inline function argument
```
I do not know if I should ignore them when writing the syntax or to create the rules necessary for these optional, but functionally useful parentheses

**Comment by ahlinc** [comment]
Usually parenthes are helpful thing cause these allow to express part of expressions as subtrees in AST and build correct evalutions based right on AST. From the other hand if you need the parser to just finding correct positions for some things in a text, such simplification like ignoring parenthes may to have sense. Also you need to think what would be in case of more complex expressions and would it be possible to parse consistently with ignored parenthes.

Pros:
 - Simplification in the grammar definition.
 - Faster parsing.
 - Simpler AST trees.

 Cons:
- No ability for correct AST driven evalutions.
- No errors detection on parser level, for ex. unbalanced parenthes.
- Not easy to just start process parenthes later. This may require to rework many parts of the grammar definition.

**Reply by katawful** [reply]
So really I can't win either way? I guess I'll have to think about if I want to implement parentheses or not, thanks for the suggestion

**Reply by ahlinc** [reply]
IMO the correct parser is the win cause it gives freedom for later AST processing.
> Faster parsing.

This is not significant pros, cause tree-sitter is fast enough.

> Simpler AST trees.

This is also achievable with tree-sitter's ability for [inlining](https://tree-sitter.github.io/tree-sitter/creating-parsers#the-grammar-dsl) or [hiding](https://tree-sitter.github.io/tree-sitter/creating-parsers#hiding-rules) rules.

**Comment by maxbrunsfeld** [comment, marked as answer]
I'm not totally sure that I'm understanding your question, but I would say that *if* your language allows parentheses, you'll need to include them in your grammar.
